solo-io/gloo
# Split webhook validation

# Description
Split the ValidatingWebhookConfiguration into 2 webhooks to allow different FailurePolicies for gloo and non-gloo resources.

## Code changes
None - changes are Helm, tests, and docs

## CI changes
Added `ValidationSplitWebhook` tests to `TestValidationStrict` tests

## Docs changes
* Helm values documentation updated
* https://github.com/solo-io/gloo/blob/75961241cc1982e62eace5b539aa8e5ed7d3837d/projects/gateway/pkg/services/k8sadmission/README.md updated

# Context
Having changes to kubernetes core objects such as secrets and namespaces go through the Gloo Service's validating webhook can prevent changes unrelated to any gloo configuration from being applied if the FailurePolicy is `Fail`. By splitting the webhooks and allowing different FailurePolicies to be set, this can be avoided.


## Testing steps
* Create cluster and build images:
```
./ci/kind/setup-kind.sh
```
* Install with the gloo FailurePolicy of `Fail` and kubes FailurePolicy of `Ignore`
```
 helm upgrade --install gloo _test/gloo-${VERSION}.tgz --create-namespace --namespace gloo-system --set-string gloo.deployment.image.tag=${VERSION} --set discovery.deployment.enablePodSecurityContext=true --set discovery.enabled=true --set gateway.validation.failurePolicy="Fail" --set gateway.validation.kubeCoreFailurePolicy="Ignore" --set gateway.validation.alwaysAcceptResources=false --set gateway.validation.allowWarnings=false --set gateway.validation.alwaysAcceptResources=false --set gateway.validation.allowWarnings=false
```
* Create an upstream (Gloo resource) and a secret (Kube resource)
```
glooctl create upstream static --static-hosts jsonplaceholder.typicode.com:80 --name json-upstream -n gloo-system
glooctl create secret apikey infra-apikey --apikey doesntmatter  
```
* Scale down the Gloo deployment:
```
kubectl scale --replicas=0 deployment/gloo -n gloo-system
```
* Try to delete the secret (should succeed)
```
 kubectl delete secret infra-apikey -n gloo-system
```
* Try to delete the upstream (should fail)
```
delete upstream -n gloo-system json-upstream
```
* Scale back up
```
kubectl scale --replicas=1 deployment/gloo -n gloo-system
```
* Update the configuration to set the gloo webhook failurePolicy to `Ignore` and the kube webhook failurePolicy to `Fail`
```
 helm upgrade --install gloo _test/gloo-${VERSION}.tgz --namespace gloo-system --set gateway.validation.failurePolicy="Ignore" --set gateway.validation.kubeCoreFailurePolicy="Fail"
```
* Recreate the deleted secret:
```
glooctl create secret apikey infra-apikey --apikey doesntmatter
```
* Scale down the Gloo deployment:
```
kubectl scale --replicas=0 deployment/gloo -n gloo-system
```
* Try to delete the secret (should fail)
```
 kubectl delete secret infra-apikey -n gloo-system
```
* Try to delete the upstream (should succeed)
```
delete upstream -n gloo-system json-upstream
```


# Checklist:
